### PR TITLE
Create Fleury's_Algorithm.py

### PR DESCRIPTION
**Pull request by DasithEdirisinghe**
### **Describe your change:**

* [x] Add an algorithm?
* [ ] Fix a bug or typo in an existing algorithm?
* [ ] Documentation change?

### **Checklist:**
* [x] I have read [CONTRIBUTING.md](https://github.com/TheAlgorithms/Python/blob/master/CONTRIBUTING.md).
* [x] This pull request is all my own work -- I have not plagiarized.
* [x] I know that pull requests will not be merged if they fail the automated tests.
* [x] This PR only changes one algorithm file.  To ease review, please open separate PRs for separate algorithms.
* [x] All new Python files are placed inside an existing directory.
* [x] All filenames are in all lowercase characters with no spaces or dashes.
* [x] All functions and variable names follow Python naming conventions.
* [x] All function parameters and return values are annotated with Python [type hints](https://docs.python.org/3/library/typing.html).
* [x] All functions have [doctests](https://docs.python.org/3/library/doctest.html) that pass the automated testing.
* [ ] All new algorithms have a URL in its comments that points to Wikipedia or other similar explanation.
* [ ] If this pull request resolves one or more open issues then the commit message contains `Fixes: #{$ISSUE_NO}`.

### CODE REVIEW

**Comment by DasithEdirisinghe** [comment]
@TravisBuddy what is the log. is there any link?

**Comment by DasithEdirisinghe** [comment]
@cclauss could you please help me.what is the error.I am new to this kind of issues.

**Comment by cclauss** [comment]
https://github.com/TheAlgorithms/Python/blob/master/Travis_CI_tests_are_failing.md

**Comment by DasithEdirisinghe** [comment]
@cclauss I cant find the issue as the code is locally working properly.What is the issue.could you please help me?

File: graphs/fleurys_algorithm.py [hunk 1]
```diff
@@ -0,0 +1,110 @@
+from collections import defaultdict
+
+
+class Graph:
+
```
Comment (cclauss) [inline graphs/fleurys_algorithm.py:5]:
```suggestion
```

Comment (cclauss) [reply graphs/fleurys_algorithm.py:5]: Let's get the tests to be green.

File: graphs/fleurys_algorithm.py [hunk 2]
```diff
@@ -0,0 +1,110 @@
...
+    def __init__(self, vertices):
+        self.v = vertices
+        self.graph = defaultdict(list)
+
+    def addEdge(self, u, v):
```
Comment (cclauss) [inline graphs/fleurys_algorithm.py:9]:
```suggestion
    def add_edge(self, universal: int, victorious: float) -> list[int]:
```
Don't force your reader to guess.  Type hints are required in this repo so the caller knows what datatypes the functions expect and what datatype they return.

Single -letter variable names for function parameters is the way that programs were written in the 1950's and 60's.  In modern times, we no longer force users to guess what variables mean.  Instead, we give them self-documenting names.

File: graphs/fleurys_algorithm.py [hunk 3]
```diff
@@ -0,0 +1,110 @@
...
+        """
+        Function to add edges to the graph
+        """
+        self.graph[u].append(v)
+        self.graph[v].append(u)
+
+    def rmvEdge(self, u, v):
```
Comment (cclauss) [inline graphs/fleurys_algorithm.py:16]:
```suggestion
    def rmv_edge(self, u, v):
```
Real variable names and type hints, please.

File: graphs/fleurys_algorithm.py [hunk 4]
```diff
@@ -0,0 +1,110 @@
...
+        Remove u-v edge from the graph
+        """
+        for index, key in enumerate(self.graph[u]):
+            if key == v:
+                self.graph[u].pop(index)
+        for index, key in enumerate(self.graph[v]):
+            if key == u:
+                self.graph[v].pop(index)
+
+    def dfsCount(self, v, visited):
```
Comment (cclauss) [inline graphs/fleurys_algorithm.py:27]:
```suggestion
    def dfs_count(self, v, visited):
```
Real variable names and type hints, please.

File: graphs/fleurys_algorithm.py [hunk 5]
```diff
@@ -0,0 +1,110 @@
...
+            visited = [False] * self.v
+            count1 = self.dfsCount(u, visited)
+
+            self.rmvEdge(u, v)
+            visited = [False] * self.v
+            count2 = self.dfsCount(u, visited)
+
+            self.addEdge(u, v)
+
+            return False if count1 > count2 else True
```
Comment (cclauss) [inline graphs/fleurys_algorithm.py:56]:
```suggestion
            return count1 <= count2
```

File: graphs/fleurys_algorithm.py [hunk 6]
```diff
@@ -0,0 +1,110 @@
...
+
+    def isValidNextEdge(self, u, v):
+        """
+        The function to check if edge u-v can be considered as next edge in
+         Euler Tour
+        """
+
+        if len(self.graph[u]) == 1:
+            return True
+        else:
```
Comment (cclauss) [inline graphs/fleurys_algorithm.py:46]:
```suggestion

```
Dedent the rest of the function.  The test above is a fast-fail.

File: graphs/fleurys_algorithm.py [hunk 7]
```diff
@@ -0,0 +1,110 @@
...
+    g3 = Graph(5)
+    g3.addEdge(1, 0)
+    g3.addEdge(0, 2)
+    g3.addEdge(2, 1)
+    g3.addEdge(0, 3)
+    g3.addEdge(3, 4)
+    g3.addEdge(3, 2)
+    g3.addEdge(3, 1)
+    g3.addEdge(2, 4)
+    g3.printEulerTour()
```
Comment (cclauss) [inline graphs/fleurys_algorithm.py:109]:
```suggestion
    graph = Graph(4)
    graph.addEdge(0, 1)
    graph.addEdge(0, 2)
    graph.addEdge(1, 2)
    graph.addEdge(2, 3)
    graph.printEulerTour()

    graph = Graph(3)
    graph.addEdge(0, 1)
    graph.addEdge(1, 2)
    graph.addEdge(2, 0)
    graph.printEulerTour()

    graph = Graph(5)
    graph.addEdge(1, 0)
    graph.addEdge(0, 2)
    graph.addEdge(2, 1)
    graph.addEdge(0, 3)
    graph.addEdge(3, 4)
    graph.addEdge(3, 2)
    graph.addEdge(3, 1)
    graph.addEdge(2, 4)
    graph.printEulerTour()
```

**Review by cclauss — CHANGES_REQUESTED** [review]
All functions need doctests, type hints on function parameters and return vaules, real variable names.

File: graphs/fleurys_algorithm.py (hunk 1 above)
Comment (DasithEdirisinghe) [reply graphs/fleurys_algorithm.py]: @cclauss thanks mate.One empty line make the difference.So how is this pull request is labeled as hacktoberfest-accepted

File: graphs/fleurys_algorithm.py (hunk 2 above)
Comment (DasithEdirisinghe) [reply graphs/fleurys_algorithm.py:9]: okay thank you

File: graphs/fleurys_algorithm.py [hunk 8]
```diff
@@ -0,0 +1,109 @@
...
+
+            return count1 <= count2
+
+    def print_euler_util(self, u) -> None:
+        """
+        To print the elular tour
+        """
+        for v in self.graph[u]:
+            if self.is_valid_next_edge(u, v):
+                print("%d-%d " % (u, v)),
```
Comment (cclauss) [inline graphs/fleurys_algorithm.py:64]:
```suggestion
                print(f"{u}-{v}"),
```
f-strings as discussed in CONTRIBUTING.md.

File: graphs/fleurys_algorithm.py [hunk 9]
```diff
@@ -0,0 +1,109 @@
...
+        self.graph = defaultdict(list)
+
+    def add_edge(self, u, v) -> None:
+        """
+        Function to add edges to the graph
+        """
+        self.graph[u].append(v)
+        self.graph[v].append(u)
+
+    def rmv_edge(self, u, v) -> None:
```
Comment (cclauss) [inline graphs/fleurys_algorithm.py:16]:
```suggestion
    def remove_edge(self, u, v) -> None:
```

File: graphs/fleurys_algorithm.py [hunk 10]
```diff
@@ -0,0 +1,109 @@
...
+        Remove u-v edge from the graph
+        """
+        for index, key in enumerate(self.graph[u]):
+            if key == v:
+                self.graph[u].pop(index)
+        for index, key in enumerate(self.graph[v]):
+            if key == u:
+                self.graph[v].pop(index)
+
+    def dfs_count(self, v, visited) -> int:
```
Comment (cclauss) [inline graphs/fleurys_algorithm.py:27]:
```suggestion
    def depth_first_search_count(self, v, visited) -> int:
```

File: graphs/fleurys_algorithm.py [hunk 11]
```diff
@@ -0,0 +1,109 @@
...
+                print("%d-%d " % (u, v)),
+                self.rmv_edge(u, v)
+                self.print_euler_util(v)
+
+    def print_euler_tour(self) -> None:
+        """
+        The main function that print Eulerian Trail. It first finds an odd
+        degree vertex (if there is any) and then calls printEulerUtil()
+        to print the path
+        """
```
Comment (cclauss) [inline graphs/fleurys_algorithm.py:73]:
```suggestion

        >>> graph = Graph(1)
        >>> graph.print_euler_tour()
        
        >>> graph.add_edge(0, 1)
        >>> graph.print_euler_tour()
        0-1
        >>> graph.add_edge(0, 1)
        >>> graph.print_euler_tour()
        
        """
```
Let's add doctests.

**Review by cclauss — CHANGES_REQUESTED** [review]

**Review by jessica02200 — APPROVED** [review]

**Review by jessica02200 — APPROVED** [review]

**Comment by poyea** [comment]
Remember to add the hacktoberfest tag in your fork repo.

**Comment by DasithEdirisinghe** [comment]
> Remember to add the hacktoberfest tag in your fork repo.

How to add that tag.could you please help me

**Comment by dhruvmanila** [comment]
@algorithms-keeper review

**Pull request closed without merging**